src/hotspot/share/opto/predicates.cpp

```diff
@@ -1135,3 +1170,80 @@ void UpdateStrideForAssertionPredicates::connect_initialized_assertion_predicate
     _phase->replace_control(new_control_out, initialized_assertion_predicate_success_proj);
   }
 }
+
+// Do the following to find and eliminate useless Parse and Template Assertion Predicates:
+// 1. Mark all Parse and Template Assertion Predicates "maybe useful".
+// 2. Walk through the loop tree and iterate over all Predicates above each loop head. All found Parse and Template
+//    Assertion Predicates are marked "useful".
+// 3. Those Parse and Template Assertion Predicates that are still marked "maybe useful" are now marked "useless" and
+//    removed in the next round of IGVN.
+//
+// Note that we only mark Predicates useless and not actually replace them now. We leave this work for IGVN which is
+// better suited for this kind of graph surgery. We also not want to replace conditions with a constant to avoid
+// interference with Predicate matching code when iterating through them.
+void EliminateUselessPredicates::eliminate() const {
+  mark_all_predicates_non_useful();
+  if (C->has_loops()) {
+    mark_loop_associated_predicates_useful();
+  }
+  mark_non_useful_predicates_useless();
+}
```

Comment:
I would replace `non_useful` to `maybe_useful` to keep it consistent with the comments above.